File: src/main/java/com/refinedmods/refinedstorage/screen/grid/view/BaseGridView.java
package com.refinedmods.refinedstorage.screen.grid.view;

import com.refinedmods.refinedstorage.api.network.grid.IGrid;
import com.refinedmods.refinedstorage.screen.grid.GridScreen;
import com.refinedmods.refinedstorage.screen.grid.filtering.GridFilterParser;
import com.refinedmods.refinedstorage.screen.grid.sorting.IGridSorter;
import com.refinedmods.refinedstorage.screen.grid.sorting.SortingDirection;
import com.refinedmods.refinedstorage.screen.grid.stack.IGridStack;

import javax.annotation.Nullable;
import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class BaseGridView implements IGridView {
    private final GridScreen screen;
    private boolean canCraft;

    private final IGridSorter defaultSorter;
    private final List<IGridSorter> sorters;

    private List<IGridStack> stacks = new ArrayList<>();
    protected final Map<UUID, IGridStack> map = new HashMap<>();

    protected BaseGridView(GridScreen screen, IGridSorter defaultSorter, List<IGridSorter> sorters) {
        this.screen = screen;
        this.defaultSorter = defaultSorter;
        this.sorters = sorters;
    }

    @Override
    public List<IGridStack> getStacks() {
        return stacks;
    }

    @Override
    public Collection<IGridStack> getAllStacks() {
        return map.values();
    }

    @Nullable
    @Override
    public IGridStack get(UUID id) {
        return map.get(id);
    }

    @Override
    public void sort() {
        if (!screen.canSort()) {
            return;
        }

        if (screen.getGrid().isGridActive()) {
            this.stacks = map.values().stream()
                    .filter(getActiveFilters())
                    .sorted(getActiveSort())
                    .collect(Collectors.toList());
        } else {
            this.stacks = Collections.emptyList();
        }

        this.screen.updateScrollbar();
    }

    private Comparator<IGridStack> getActiveSort() {
        IGrid grid = screen.getGrid();
        SortingDirection sortingDirection = grid.getSortingDirection() == IGrid.SORTING_DIRECTION_DESCENDING ? SortingDirection.DESCENDING : SortingDirection.ASCENDING;
        return Stream.concat(Stream.of(defaultSorter), sorters.stream().filter(s -> s.isApplicable(grid)))
                .map(sorter -> (Comparator<IGridStack>) (o1, o2) -> sorter.compare(o1, o2, sortingDirection))
                .reduce((l, r) -> r.thenComparing(l))
                .orElseThrow(IllegalStateException::new);  // There is at least 1 value in the stream (i.e. defaultSorter)
    }

    private Predicate<IGridStack> getActiveFilters() {
        IGrid grid = screen.getGrid();

        Predicate<IGridStack> filters = GridFilterParser.getFilters(
                grid,
                screen.getSearchFieldText(),
                (grid.getTabSelected() >= 0 && grid.getTabSelected() < grid.getTabs().size()) ? grid.getTabs().get(grid.getTabSelected()).getFilters() : grid.getFilters()
        );

        if (screen.getGrid().getViewType() != IGrid.VIEW_TYPE_CRAFTABLES) {
            return stack -> {
                // If this is a crafting stack,
                // and there is a regular matching stack in the view too,
                // and we aren't in "view only craftables" mode,
                // we don't want the duplicate stacks and we will remove this stack.
                if (stack.isCraftable() &&
                        stack.getOtherId() != null &&
                        map.containsKey(stack.getOtherId())) {
                    return false;
                }

                return filters.test(stack);
            };
        } else {
            return filters;
        }
    }

    @Override
    public void setCanCraft(boolean canCraft) {
        this.canCraft = canCraft;
    }

    @Override
    public boolean canCraft() {
        return canCraft;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/screen/grid/view/BaseGridView.kt
package com.refinedmods.refinedstorage.screen.grid.view

import com.refinedmods.refinedstorage.api.network.grid.IGrid
import com.refinedmods.refinedstorage.screen.grid.GridScreen
import com.refinedmods.refinedstorage.screen.grid.filtering.GridFilterParser
import com.refinedmods.refinedstorage.screen.grid.sorting.IGridSorter
import com.refinedmods.refinedstorage.screen.grid.sorting.SortingDirection
import com.refinedmods.refinedstorage.screen.grid.stack.IGridStack
import java.util.*

abstract class BaseGridView(private val screen: GridScreen, private val defaultSorter: IGridSorter, private val sorters: List<IGridSorter>) : IGridView {
    private var canCraft = false
    private var stacks: List<IGridStack?> = ArrayList()
    protected val map: Map<UUID, IGridStack?> = HashMap()
    override fun getStacks(): List<IGridStack?> {
        return stacks
    }

    @Nullable
    override fun get(id: UUID?): IGridStack? {
        return map[id]
    }

    override fun sort() {
        if (!screen.canSort()) {
            return
        }
        val stacks: MutableList<IGridStack?> = ArrayList()
        if (screen.grid.isGridActive) {
            stacks.addAll(map.values)
            val grid = screen.grid
            val filters = GridFilterParser.getFilters(
                    grid,
                    screen.searchFieldText,
                    if (grid!!.tabSelected >= 0 && grid.tabSelected < grid.tabs!!.size) grid.tabs!![grid.tabSelected]!!.filters else grid.filters
            )
            stacks.removeIf { stack: IGridStack? ->
                // If this is a crafting stack,
                // and there is a regular matching stack in the view too,
                // and we aren't in "view only craftables" mode,
                // we don't want the duplicate stacks and we will remove this stack.
                if (screen.grid.viewType != IGrid.VIEW_TYPE_CRAFTABLES &&
                        stack!!.isCraftable && stack.otherId != null &&
                        map.containsKey(stack.otherId)) {
                    return@removeIf true
                }
                for (filter in filters!!) {
                    if (!filter!!.test(stack)) {
                        return@removeIf true
                    }
                }
                false
            }
            val sortingDirection = if (grid.sortingDirection == IGrid.SORTING_DIRECTION_DESCENDING) SortingDirection.DESCENDING else SortingDirection.ASCENDING
            stacks.sort(java.util.Comparator<IGridStack> { left: IGridStack, right: IGridStack -> defaultSorter.compare(left, right, sortingDirection) })
            for (sorter in sorters) {
                if (sorter.isApplicable(grid)) {
                    stacks.sort(java.util.Comparator<IGridStack> { left: IGridStack, right: IGridStack -> sorter.compare(left, right, sortingDirection) })
                }
            }
        }
        this.stacks = stacks
        screen.updateScrollbar()
    }

    override fun setCanCraft(canCraft: Boolean) {
        this.canCraft = canCraft
    }

    override fun canCraft(): Boolean {
        return canCraft
    }
}